File: crawlers/biospectator.go
package crawlers

import (
	"fmt"
	"net/http"

	"flashnews/models"
	"flashnews/utils"

	"github.com/PuerkitoBio/goquery"
)

const bioSpectatorCommonURL = "http://www.biospectator.com/section/section_list.php?MID=10000"
const bioSpectatorItemURL = "http://www.biospectator.com"

type BioSpectator struct{}

func (c BioSpectator) GetName() string {
	return "biospectator"
}

func (c BioSpectator) GetList(number int) ([]models.NewsItem, error) {
	// Number
	var _number int
	if number > 15 || number < 1 {
		_number = 15
	} else {
		_number = number
	}
	result := make([]models.NewsItem, _number)

	// Request
	req, err := http.Get(bioSpectatorCommonURL)
	if err != nil {
		return result, err
	}
	defer req.Body.Close()

	if req.StatusCode != 200 {
		return result, fmt.Errorf("[ERROR] Request Status Code : %d, %s", req.StatusCode, req.Status)
	}

	// HTML Load
	html, err := goquery.NewDocumentFromReader(req.Body)
	if err != nil {
		return result, err
	}

	// Parsing
	wrapper := html.Find("div.article_list")
	items := wrapper.Find("li")
	items.Each(func(i int, sel *goquery.Selection) {
		if i >= _number {
			return
		}

		strong := sel.Find("strong.article_tit")
		aTag := strong.Find("a")
		href, ok := aTag.Attr("href")
		if !ok {
			result[i] = models.NewsItem{}
			return
		}
		url := bioSpectatorItemURL + href

		date := sel.Find("span.date").Text()
		title := utils.TrimAll(aTag.Text())
		if err != nil {
			result[i] = models.NewsItem{}
			return
		}

		result[i] = models.NewsItem{
			Title:    title,
			URL:      url,
			Contents: "",
			Datetime: date,
		}
	})

	return result, nil
}

func (c BioSpectator) GetContents(item *models.NewsItem) error {
	// Request
	req, err := http.Get(item.URL)
	if err != nil {
		return err
	}
	defer req.Body.Close()

	if req.StatusCode != 200 {
		return fmt.Errorf("[ERROR] Request Status Code : %d, %s", req.StatusCode, req.Status)
	}

	// HTML Load
	html, err := goquery.NewDocumentFromReader(req.Body)
	if err != nil {
		return err
	}

	// Parsing
	wrapper := html.Find("div.article_view")
	contents := ""
	wrapper.Find("p").Each(func(idx int, sel *goquery.Selection) {
		if idx > 0 {
			contents += utils.TrimAll(sel.Text())
		}
	})

	item.Contents = contents
	return nil
}

File: pys/biospectator.py
import requests
from bs4 import BeautifulSoup

url = 'http://www.biospectator.com/section/section_list.php?MID=10000'

req = requests.get(url)
bs = BeautifulSoup(req.text, 'lxml')

wrapper = bs.find('div', class_='article_list')
items = wrapper.find_all('li')

for item in items[:1]:
    span = item.find('strong', class_='article_tit')
    a_tag = span.find('a')
    title = a_tag.get_text().strip()
    href = 'http://www.biospectator.com' + a_tag['href']
    date = item.find('span', class_='date').get_text().strip()

    req2 = requests.get(href)
    bs2 = BeautifulSoup(req2.content, 'lxml')
    wrapper2 = bs2.find('div', class_='article_view')

    p_tags = wrapper2.find_all('p')[1:]
    contents = ''
    for p in p_tags:
        if p.has_attr('class'):
            if p['class'][0] == 'photo_caption':
                continue
        contents = contents + p.get_text()
    print(title, href, date, contents)